azure/ConsiderationsForServiceDesign.md

```diff
@@ -229,6 +229,33 @@ An important distinction between RELO and status monitor LROs is that there is a
 status monitor LRO, whereas the status of all RELO operations is combined into the status of the resource.
 So status monitor LROs are "one-to-one" with their operation status, whereas RELO-style LROs are "many-to-one".
 
+## Errors
+One of the most important parts of service design is also one of the most overlooked.  The errors returned by your service are a critical part of your developer experience.  Your service and your customer's application together form a distributed system.  Errors are inevitable, but well-designed errors can help you avoid costly customer support incidents by empowering customers to self-diagnose problems.
+
+First, you should always try to design errors out of existence if possible.  You'll get a lot of this for free by following the Guidelines.  Some examples include:
+- Idempotent APIs solve a whole class of network issues where customers have no idea how to proceed if they send a request to the service but never get a response.
+- Accessing resources from multiple microservices can quickly lead to complex race conditions so conditional requests provide optimistic concurrency for safe usage.
+- Reframing the purpose of an API can obviate some errors.  This is most often specific to your operations, but an example from the Guidelines is thinking about `DELETE`s as _"ensure no resource at this location exists"_ so they can return an easier to use `204` instead of _"delete this exact resource instance"_ which would fail with a non-idempotent `404`.
```

Comment:
Yes, spec lawyers will point out that DELETE in HTTP is defined as having to be idempotent. DELETEing the resource more than once doesn't change the state on the server, which is what matters for idempotency purposes. 

We can point out that you can combine conditional requests/If-Match: * and DELETE for clients that really really want to know that the DELETE actually happened. If we want an annotated version of the guidelines.

Reply:
This is an area where our default client experience for a 204 + 404 DELETE feels very non-idempotent.  Getting a 204 response on your first request and an exception that hijacks your control flow on the second feels like you're seeing different results for the same request.  Removed the `non-idempotent` though to avoid confusing folks and/or inspiring this debate again in the future.  :smile: